### ISSUE TITLE
iptables state needs to keep ordering of flags

### ISSUE THREAD

**Issue by c4urself** [saltstack/salt#24226]
[Version info]
```
root@ip-10-0-100-130:~# salt-call --versions-report
                  Salt: 2015.5.0
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2015.5.0+ds-1trusty1
```

The problem I'm having is that ordering is sometimes important for iptables' flags, the following state:
```
{% for route_name, route in pillar['vpn']['snat_routes'].items() %}
route_to_vpn_{{route_name}}:
  iptables.append:
    - table: nat
    - chain: POSTROUTING
    - source: 172.16.0.0/24
    - destination: {{route.addr}}
    - jump: SNAT
    - "to-source": {{ grains['ec2_local-ipv4'] }}
```

Is implemented with this command:

`/sbin/iptables --wait -t nat -A POSTROUTING  --source 172.16.0.0/24 --to-source 10.0.100.130 --destination 1.2.3.4 --jump SNAT`

The problem is that `--to-source` *must* appear after the `--jump` flag. Changing the command makes it work correctly:
` /sbin/iptables --wait -t nat -A POSTROUTING  --source 172.16.0.0/24 --destination 1.2.3.4 --jump SNAT --to-source 10.0.100.130`

**Comment by techhat** [comment]
This is resolved with #23308. @jfindlay, can we backport that?

**Comment by rallytime** [comment]
@techhat I've backported it in #24232. There was a pretty decent merge conflict, because of some refactoring in the iptables module. After looking at the changes in #23308, #23241, and #20593 I *think* I resolved it correctly, but I am not certain.

@c4urself @thusoy @garethgreenaway @dmyerscough - Can you guys look that over and test it if possible and make sure I didn't break anything with that back-port?

**Comment by thusoy** [comment]
@rallytime It seems something got broken, the commands issued are missing a space a between the jump and the jump arguments (resulting in rules like `--jump LOG--log-prefix "myprefix: "`)

[This](https://github.com/saltstack/salt/blob/2733f66449e8014fa60dbcdc750c166d9a817983/salt/modules/iptables.py#L392):
```python
for item in after_jump:
    rule += item
```
breaks, because the after_jump arguments do not end with a space. You can fix this by either changing those two lines to `rule += ' '.join(after_jump)`, or backport my [initial refactoring](https://github.com/saltstack/salt/pull/20593), which also includes tests that would have caught this.

**Comment by rallytime** [comment]
@thusoy Thanks for testing and for the tip. I've added the `join` to add the space in the commands. Usually we try not to backport such big refactors to older, stable branches. Hopefully that little change will do just the trick! Otherwise, I can look at making an exception to backporting the whole refactor.

**Comment by thusoy** [comment]
@rallytime Sure, makes sense. 

See the comment I gave in the PR you created, with that change it works for my servers.

**Comment by thusoy** [comment]
This was fixed by e2d1606b19c1242187991c2a5310ab93cc756bee and can be closed.

**Comment by rallytime** [comment]
Awesome! Thanks for the follow up and reminder @thusoy!
